core/src/apps/bitcoin/verify_message.py

```diff
@@ -30,17 +67,16 @@ async def verify_message(ctx: wire.Context, msg: VerifyMessage) -> Success:
 
     digest = message_digest(coin, message)
 
+    script_type = address_to_script_type(address, coin)
     recid = signature[0]
     if 27 <= recid <= 34:
-        # p2pkh
-        script_type = InputScriptType.SPENDADDRESS
-    elif 35 <= recid <= 38:
+        # p2pkh or no script type provided
+        pass  # use the script type from the address
+    elif 35 <= recid <= 38 and script_type == InputScriptType.SPENDP2SHWITNESS:
```

Comment:
will `verify_recover()` fail if we don't tweak the signature?

Reply:
Yes, it would fail. Why is this important?

Reply:
Because this is the only recid <-> script-type crosscheck. If it wouldn't fail, then the "and script_type == something" checks are pointless.

Reply:
I don't think they would be pointless. If the `and script_type == ...` condition fails, then we end up raising a `ProcessError` on line 82, i.e. never reach `verify_recover()`.

Reply:
right, i somehow completely missed that. thanks.